### PR TITLE
Add more packets (and minor fixes)

### PR DESCRIPTION
**Pull request by TheSnoozer**
First of all: thanks for this awesome project!
I decided to play around with it and added some packets to the python implementation (including respawning which seems to be the first actual client command that is not being dealt on server side).

I also added some workaround that deals with faulty compression that for some reason gets introduced when using a specific plugin that supports the play to join a server with multiple versions. Not 100% what happens but based on the packets the server claims to have compression enabled but in fact it is not.

I have tested with the following:
* git-Spigot-21fe707-e1ebe52 (MC: 1.8.8) (Implementing API version 1.8.8-R0.1-SNAPSHOT)
* git-Spigot-c6871e2-0cd0397 (MC: 1.9.4) (Implementing API version 1.9.4-R0.1-SNAPSHOT)
* git-Spigot-de459a2-51263e9 (MC: 1.10.2) (Implementing API version 1.10.2-R0.1-SNAPSHOT)
* git-Spigot-6f7aabf-c8ff651 (MC: 1.11) (Implementing API version 1.11-R0.1-SNAPSHOT)
* git-Spigot-596221b-9a1fc1e (MC: 1.12) (Implementing API version 1.12-R0.1-SNAPSHOT)

even though I have tested it with almost every major release I can't guarantee its bug free...

### CODE REVIEW

**Comment by ammaraskar** [comment]
Thanks a lot for your work! Will review this when I'm free, and I'd appreciate a look from you as well @joodicator

File: minecraft/networking/connection.py [hunk 1]
```diff
@@ -436,14 +436,23 @@ def read_packet(self, stream):
...
-                if decompressed_size > 0:
-                    decompressed_packet = decompress(packet_data.read())
-                    assert len(decompressed_packet) == decompressed_size, \
-                        'decompressed length %d, but expected %d' % \
-                        (len(decompressed_packet), decompressed_size)
-                    packet_data.reset()
-                    packet_data.send(decompressed_packet)
-                    packet_data.reset_cursor()
+                rawPacketData = None
+                try:
```
Comment (ammaraskar) [inline minecraft/networking/connection.py]: I'm not convinced this is the right way to handle this, can you provide more information about when this problem happens? What server with what plugin configuration?

File: minecraft/networking/types.py [hunk 2]
```diff
@@ -213,3 +223,31 @@ def read(file_object):
...
+        if y >= pow(2, 11):
+            y -= pow(2, 12)
+
+        if z >= pow(2, 25):
+            z -= pow(2, 26)
+
+        return {'x': x, 'y': y, 'z': z}
+
+    @staticmethod
+    def send(value, socket):
```
Comment (ammaraskar) [inline minecraft/networking/types.py]: I would suggest changing this to pass either the x, y, z as arguments or passing a dict containing x, y, z. The types classes are supposed to abstract away the python types and sending them over the network and calling an encode method before sending breaks that a little.

File: minecraft/networking/packets.py [hunk 3]
```diff
@@ -727,6 +727,399 @@ def __str__(self):
...
+
+    packet_name = 'block change'
+
+    def read(self, file_object):
+        self.location = Position.read(file_object)
+        blockData = VarInt.read(file_object)
+        self.blockId = (blockData >> 4)
+        self.blockMeta = (blockData & 0xF)
+
+    def write(self, socket, compression_threshold=None):
```
Comment (ammaraskar) [inline minecraft/networking/packets.py]: I would suggest implementing this so this is easier to test. See this commit for an example: https://github.com/ammaraskar/pyCraft/commit/d686b6487f8936fb6ed390386390632502b718b9

File: minecraft/networking/packets.py [hunk 4]
```diff
@@ -727,6 +727,399 @@ def __str__(self):
...
+                90: ClientSpawnObject.Type_Fishing_Hook,
+                91: ClientSpawnObject.Type_Spectral_Arrow,
+                93: ClientSpawnObject.Type_Dragon_Fireball
+            }.get(type_id)
+            if subcls is None:
+                raise ValueError("Unknown type ID: %s."
+                                 % type_id)
+            return subcls
+
+    class Type_Boat(Type):
```
Comment (ammaraskar) [inline minecraft/networking/packets.py]: I think you followed the example from the `PlayerListItemPacket` for this, the reason it has this subclassing is because the packet varies depending on the type of Action. In this case, the type doesn't vary so you can forego this and simply introduce an enum like thing that looks like:

```python
class EntityType:
    BOAT = 1
    ITEM_STACK = 2
    # ...

    by_id = {id: entity for (entity, id) in EntityType.__dict__.items() if entity.isupper()}

    @staticmethod
    def get_entitity_type_by_id(id):
        return self.by_id[id]
```

**Comment by ammaraskar** [comment]
@joodicator So looking at this especially, I think `packets.py` is getting quite crowded. We should work out a way of organizing out the more complicated packets so we don't end up with a lot of code to handle all the cross version changes in one place.

**Comment by TheSnoozer** [comment]
@ammaraskar thanks for the feedback / input and I'll def look into this further.

What I also would suggest is some sort of common pattern for client bound packets and server bound packets. 
Some are named like  ``$name_Clientbound`` (e.g. KeepAlivePacketClientbound); 
some don't have a name if they are client or server bound e.g. ChatMessagePacket or MapPacket. 
Last but not least the packets I introduced all start with ``Client$name`` or ``Server$name`` to somewhat distinguish the usecase....

File: minecraft/networking/connection.py (hunk 1 above)
Comment (TheSnoozer) [inline minecraft/networking/connection.py]: Unfortunately I don't know anything in specific about the plugins used nor the server version (afaik they use 1.8.8).
In general this faulty behaviour can be reproduced on play(.)saicopvp(.)com

I did some further investigation and the server claims to support the following protocol versions:
[47, 107, 108, 109, 110, 210, 301, 302, 303, 304, 305, 306, 307, 308, 309, 310, 311, 312, 313, 314, 315, 316, 317, 318, 319, 320, 321, 322, 323, 324, 325, 326, 327, 328, 329, 330, 331, 332, 333, 334, 335, 336, 337, 338]

However when connecting with ``self.context.protocol_version = max(self.allowed_proto_versions)`` the client doesn't do anything.
After some debugging I found out that the server sends the following: ``0x1A 0x1A DisconnectPacketPlayState {'json_data': '"Unsupported protocol version 335"'}`` or similar depending on the version the client connects with.

In general this is messed up from the start already. The server claims to support this specific version and when connecting it sends a DisconnectPacketPlayState-Packet.

Regardless for my testing I adjusted the way how the version is determined by using:
``self.context.protocol_version = min(self.allowed_proto_versions)`` 

this will result in:
``  File "~/pyCraft/minecraft/networking/connection.py", line 440, in read_packet
    decompressed_packet = decompress(packet_data.read())
zlib.error: Error -5 while decompressing data: incomplete or truncated stream
``
If I add output about the `decompressed_size` its a non-zero value and dynamic for each packet. this it will try to decompress the packet and fail with the previous error.

I can't reproduce this in any way with any paper spigot / spigot server with ViaVersion.

**Comment by ammaraskar** [comment]
>@ammaraskar Perhaps we could separate the packets into sub-modules based on whether they are serverbound or clientbound, and in what connection state they belong

That sounds like a solid plan, let's pull this with consistent Packet names and then implement these style of modules.

File: minecraft/networking/types.py [hunk 5]
```diff
@@ -213,3 +223,28 @@ def read(file_object):
...
+        if x >= pow(2, 25):
+            x -= pow(2, 26)
+
+        if y >= pow(2, 11):
+            y -= pow(2, 12)
+
+        if z >= pow(2, 25):
+            z -= pow(2, 26)
+
+        return {'x': x, 'y': y, 'z': z}
```
Comment (ammaraskar) [inline minecraft/networking/types.py]: Creating a `Position` class would have the additional benefit that it contains all the fields required for `ClientExplosion.Record`, so that could simply subclass Position and pass.

**Comment by TheSnoozer** [comment]
Thanks for the feedback guys!
I'll look into those and work on them asap :-)

File: minecraft/networking/types.py (hunk 5 above)
Comment (joodicator) [inline minecraft/networking/types.py]: > Creating a Position class would have the additional benefit that it contains all the fields required for ClientExplosion.Record, so that could simply subclass Position and pass.

If that is to be done, then it might be better for the named tuple to be a common superclass of both `types.Position` and `packets.ClientExplosion.Record`, so that instances of the latter do not include `read` and `send` methods. Note that`ClientExplosion.Record` are encoded as an array of 3 bytes, while the fields in a `Position` are encoded as 26 + 26 + 12 bits packed into an unsigned long integer, so the two are best not confused.

**Comment by TheSnoozer** [comment]
I added most of the requested changes.
Please note that this is not everything yet since I still need to fix/address the faulty compression feedback.

to silence the following errors:
```
minecraft/networking/packets/__init__.py:13:1: F401 '.clientbound.handshake.get_packets as state_handshake_clientbound' imported but unused
```
I included a ``# flake8: noqa`` inside ``pyCraft/minecraft/networking/packets/__init__.py`` not sure if there is a better way to exclude specific errors with flake...

**Comment by TheSnoozer** [comment]
I added all the requested changes.
Please feel free to suggest further input.

The only thing I'm not really happy about is ``# flake8: noqa`` inside ``pyCraft/minecraft/networking/packets/__init__.py`` to silence ``imported but unused`` errors.

**Comment by ammaraskar** [comment]
A better fix for the flake8 issue would be to add
```python
from .clientbound.handshake import get_packets as state_handshake_clientbound

__all__ = ['state_handshake_clientbound', ...]
```

to show that the import is being used as an export

**Comment by joodicator** [comment]
Nice work! There are just a few packet names still containing some redundant information that I would recommend changing:

Module | Current Name | Suggested Name
---|---|---
clientbound.play | DisconnectPacketPlayState | DisconnectPacket
clientbound.play | ChatMessagePacket | ChatPacket
clientbound.play | SetCompressionPacketPlayState | SetCompressionPacket
clientbound.status | PingPacketResponse | PingResponsePacket

The last one may be arguable. It introduces more inconsistency between the old and new names, but IMO it is worth it for the new names to be more consistent among themselves. (I also considered PongPacket, which would be inline with the names now used on wiki.vg, but I don't think that's a good idea, as the `packet_name` attribute is `"ping"`, and that can't be reasonably changed).

Other than that, this seems good to go to me.

**Comment by joodicator** [comment]
Actually, on that same point, perhaps ChatMessagePacket should be left as it is, because its `packet_name` is `"chat message"`.

File: minecraft/networking/types.py [hunk 6]
```diff
@@ -4,6 +4,7 @@
 """
 import struct
 import uuid
+from collections import namedtuple
 
 
 class Type(object):
```
Comment (joodicator) [inline minecraft/networking/types.py]: Oh, and I almost forgot: the class `Type` here needs to also have `__slots__ = ()`, otherwise every instance of `Position` will have an unnecessary empty `__dict__` allocated.

**Comment by TheSnoozer** [comment]
Thanks for the input!
I have adjusted the naming and added the ``__slots__ = ()`` definition to ``Type``

**Comment by joodicator** [comment]
@TheSnoozer When implementing [this change](http://wiki.vg/index.php?title=Pre-release_protocol&oldid=13582#Global_palette) to the global palette of block state IDs used by the Block Change and Multi Block Change packets, which were added in this PR, I felt the need to make some backward-incompatible changes to these packets, [shown here](https://github.com/ammaraskar/pyCraft/commit/f492adfeffde7c95aedb407577edc249711e27bc#diff-881bd7ef264858db9252ec185bdafe6f).

The issue is that the block data can no longer be split into a block ID and a state ID just using bitwise operations, so I felt it was best to leave it as an opaque block state ID for library users to deal with as they see fit. I would be interested to know if this is a helpful change for any use cases of the Block Change and Multi Block Change packets you might have. I'd be pleased to hear any suggestions as to how the implementation might be improved, since I'm not sure my solution is the best one.

**Comment by TheSnoozer** [comment]
Hi @joodicator,
sorry to say even after reading the change on the protocol I might not fully understand what Minecraft is doing there....let me try to summarize what I understood from the issue:
In older minecraft versions Block Change packet contained a Block ID  that could be split up into tpye and some metadata using bitwise operations (type = id >> 4, meta = id & 15). However with the palette change for 1.12.2, protocol 340 this does not work anymore. I don't quite understand how the new block id assignment works. Based from the protocol the Block Ids are created in a linear fashion based off of order of assignment. Does this mean that each stone has a different block id or does this affect the "state"/"metadata"? 

For me the main use-case is to be able to determine what type of block we have. If you are interested in a specific use-case feel free to drop me a mail (git log should tell you ;-)).

**Comment by joodicator** [comment]
@TheSnoozer Each block may be represented by multiple block state IDs. For example, `acacia_door` could be represented by any block state ID from 6224 to 6287, with the 64 different values representing different combinations of upper/lower half, facing direction, open state, hinge orientation, and powered state. This is similar to how it was before, except now the only way to separate the block ID from its metadata would be to look it up in a table, and it seems that the entries in the table could change between versions as blocks are added or removed.

In your use case, if you want to check if a block is an `acacia_door` block, you presumably currently do something equivalent to checking if `blockId == 196`. Now, you would need to do something equivalent check if `6224 <= blockStateId <= 6287`. (You would, presumably, prefer to have some kind of lookup table rather than distributing magic constants throughout your code, but that is beside the point). If you wanted your program to work for versions both after and prior to protocol 347, you'd need to do it differently depending on the protocol version. If the Global Palette referred to on wiki.vg changes in future versions (which I expect it might), you'd have to deal with that on a per-version basis in your code as well.

My question is: bearing in mind how you'd have to modify your program to make it work with these changes, is there anything that could be changed about pyCraft's API that would make it less painful a change to deal with?

**Comment by TheSnoozer** [comment]
@joodicator thanks for the clarification...this makes much more sense now.
On the one hand I'd say it would be super convenient to have some sort of lookup-table in pyCraft's API available. On the other hand it feels that the client needs to be aware of the version that is being used to some extend anyways. Just FYI as of right now I actually use the had coded values for certain blocks (aka. ``if blockId == 196:``). The changes with the new version would be somewhat trivial (still a pain if i want a client supporting different versions).

In terms what feels the 'best' would be potentially a lookup (prop an enum) inside pyCraft's API that tells me what type of block we have. However I fear maintaining that for multiple different versions would be a pain in the ***.

Just for reference I originally used the MCProtocolLib (java implementation). I changed cause it was just killing my RAM. Looking at the code (https://github.com/Steveice10/MCProtocolLib/blob/master/src/main/java/com/github/steveice10/mc/protocol/packet/ingame/server/world/ServerMultiBlockChangePacket.java and https://github.com/Steveice10/MCProtocolLib/blob/master/src/main/java/com/github/steveice10/mc/protocol/data/game/world/block/BlockState.java) the implementation just gives you a blockid. That somewhat reflects the code I had written there where I build a client side wrapper to translate the hard-coded values to something more readable (enum).

Just a side node: The servers I use the client on do not support 12.X so to that extend I'm not even affected by this change (plus I connect with a hard coded version so my hard coded block ids are guaranteed not changing). Regardless of that it would be convenient to have a lookup. The benefit would be that the client code do not really to bother about different versions (since it would be reflected inside the API's lookup) plus any hard coded (somewhat unreadable magic) can be replaced with something more readable. As you said we currently don't know how the protocol will change in the feature so this would be the ideal world. 

TLDR:
- Lookup in that case would be super sweet and would allow a more generic and flexible approach with changing the protocol versions on the client side
- I use hard coded comparisons ``blockId == 196`` and could easily adapt the changes since I feel the client needs to be aware of the version to some extend

**Comment by joodicator** [comment]
@TheSnoozer I see. I agree that it would be a good feature for pyCraft to be aware of block data, and to allow library users to work with tokens like `acacia_door` in a somewhat version-independent way, instead of numbers. At the moment, though, I am limiting myself to maintaining the current level of functionality. Considering this, I'll leave the current implementation as it is. If you happen to have any more thoughts on the design before the release of Minecraft 1.13, do make them known.

**Comment by TheSnoozer** [comment]
Ok sounds fair for me. If there is anything I see fitting I'd open up a MR or create an issue.
I fully understand that your time is limited (same as mine - still have hundreds of things I want to do with this project J, but barely got any free time :( )
Anyways thanks for double checking and this great project (still loads of fun with it xD).
